hatchways/team-bamboo
# Backend Messaging - Model, Routes and Controllers

### Closes #3 

### What this PR does (required):

#### New Data Models
##### 1. Conversation -> represents a conversation between users
###### fields
- user1 -> A mongo id referencing a user's profile.
- user2 -> A mongo id referencing a user's profile.
- lastMessage -> a mongo id referencing the last message sent within the conversation.

##### 2. Message -> represents the data structure of a message between two users
###### fields
- sender -> a mongo id referencing a user's profile.
- conversationId -> a mongo id referencing the conversation
- content -> a string value containing the content of a message.
- createdAt -> a date which represents the date the message was sent.
- updatedAt -> a date which represents the last time the message was updated.
---
#### New Routes
##### 1. Route: /conversations -> used to get all conversations between two users.
##### method -> GET
###### response example:
```json
{
    "success": {
        "conversations": [
            {
                "id": "conversation id",
                "otherUser": {
                    "name": "Some User's Name",
                    "photo": "https://example.com",
                    "id": "other user's profile id"
                },
                "lastMessage": {
                    "sender": "sender's profile id",
                    "content": "Hello world!",
                    "createdAt": "2022-01-30T09:19:30.216Z",
                    "id": "last message's id"
                }
            }
        ]
    }
}
```
#### New Routes
##### 2. Route: /conversations -> creates a new conversation between two users if one doesn't already exist.
##### method -> POST
###### body
- otherUserId -> a mongo user id. Uses User model id because of the route for searching for a specific user. Cannot be the same as the user "creating" the new conversation.
###### body:
```json
{
    "otherUserId": "some user's id",
}
```
###### response example:
```json
{
    "success": {
        "conversation": {
            "id": "conversation id",
            "otherUser": {
                "name": "Some User's Name",
                "photo": "https://example.com",
                "id": "other user's profile id"
            },
        }
    }    
}
```

##### 3. Route: conversations/:id/messages -> fetches all messages with provided conversation id.
##### method -> GET
###### params
- id -> conversation id in which the message is located.
###### query
- limit -> number representing the amount of messages to fetch. Default: 30
- page -> number value. Default: 1
- sort -> enum value specific to fields relevant to the message schema. Default: "createdAt"
- order -> enum value ('desc' or 'asc') representing the order in which to return the messages. Default: 'asc'
###### response structure:
```json
{
    "success": {
        "messages": [
            {
                "sender": {
                    "_id": "id of sender",
                    "name": "Name of Sender",
                    "photo": "https://example.com"
                },
                "conversationId": "id of conversation",
                "content": "Hello world!",
                "createdAt": "2022-01-30T08:08:35.503Z",
                "updatedAt": "2022-01-30T08:08:35.503Z",
                "__v": 0,
                "id": "message id"
            }
        ]
    }
}
```
##### 4. Route: /conversations/:id/messages -> creates a new message with the provided conversation id.
##### method -> POST
###### body
- content -> a string value representing the text content of a string
###### body structure:
```json
{
    "content": "Hello world!",
}
```
###### response structure:
```json
{
    "success": {
        "message": {
            "id": "message id",
            "sender": {
                "id": "profile id of sender",
                "name": "Some Sender's Name",
                "photo": "https://example.com"
            },
            "conversationId": "conversation id",
            "content": "Hello world!",
            "createdAt": "2022-01-30T09:59:34.196Z",
        }
    }
}
```

### Any information needed to test this feature (required):
- Create two test user accounts
- Sign into the first test user account and create a message using the '/messages/send' endpoint. Make sure to provide the receiverId which is a reference to the second test user's account and the content of the message.
- Visit endpoint: /conversations -> to see the newly created conversation between test user 1 and test user 2. Store the conversation id to fetch the messages.
- Visit endpoint: /messages/:convoId -> to see all messages between test user 1 and 2.
